Added file fields for a release's tarball and checksum

bmispelon opened the pull request:
This is in preparation for https://code.djangoproject.com/ticket/35980 which will likely change the naming scheme of build files (tarballs and checksums).

Here's a summary of the changes in this PR:

* Added three new (optional) file fields to the `Release` model: `tarball`, `wheel`, and `checksum`
* Wrote a data migration to populate the file fields based on existing data, so that existing https://www.djangoproject.com/m/releases/...` links keep working (see [gist](https://gist.github.com/bmispelon/111261f716ddbfa529e5286f4b696197) for data/scripts used).
* Add some custom model validation to make the `tarball` field mandatory when a release is active (ie it has a release date). This is because the template in `download.html` assumes that all published/active releases have at least a tarball attached to them.
* Added some fieldsets on the release admin form (see screenshot) to make the form a bit neater.
* Added an `accept` attribute on the admin form file widgets to make it easier to pick the right file for each
* Removed all hardcoded filenames for tarballs and checksums across the site
* Removed custom logic to compute release file names based on date/version number (the actual file names from the `Release` model can now be used)

<details>
<summary>Screenshots</summary>

![Screenshot 2025-02-15 at 13-26-58 5 2 Change release Django site admin](https://github.com/user-attachments/assets/812c11a5-f006-4075-8d42-67d56948f91e)
</details>

I've also prepared a companion PR to the documentation of the release process over at https://github.com/django/django/pull/19178

nessita commented:
On a second look, and after some local testing, I wonder why we don't spare the new fields and just make them a property? That seems simpler and that it scales better...

nessita commented:
@bmispelon Hi! Do you think you'd have any availability to continue your work on this?

nessita commented:
@bmispelon Hi! Could you please help me understand whether you'd have availability to keep on this? Otherwise, no worries, I can take it: `setuptools` is going to be released soon with the fix so we are planning next steps to push this and other changes around it forward. Thank you!

bmispelon commented:
> @bmispelon Hi! Could you please help me understand whether you'd have availability to keep on this? Otherwise, no worries, I can take it: `setuptools` is going to be released soon with the fix so we are planning next steps to push this and other changes around it forward. Thank you!

I've picked this up again, thanks for the ping. I've rebased the branch, fixed the CI, and updated the first comment with a list of what's left for me to do.
I'm hoping to be done this week 🤞🏻  (if not, I'll ping you and we can figure out if it makes sense for you to take over)

bmispelon commented:
@nessita I think this is ready for a re-review. I've taken your advice and simplified the migration a bit so that only exceptions to the general naming rule are listed explicitly. I've always made a few quality-of-life improvements to the admin form (fieldsets, custom validation) but if you think that's going to get in the way of your work please let me know.

nessita reviewed (commented):
Thank you @bmispelon for this work, I have applied my suggestions locally and I have made other changes based on what we talked during the DSF Office Hours (pushing that now).

I haven't fixed the tests, would you have any availability to fix them? :star2:

bmispelon commented on djangoproject/templates/releases/_download_links.html:
```diff
@@ -0,0 +1,9 @@
+
+<a href="{% url 'download-redirect' release.version 'tarball' %}">
+  {{ release.tarball.name }}</a><br>
+{% if release.checksum %}
```
The `if` is there for older legacy versions that don't have a checksum, I think it has to stay otherwise those old versions will have a broken link.

nessita replied:
Good point, re-adding the `if` guard.

bmispelon commented on djangoproject/templates/releases/_download_links.html:
```diff
@@ -0,0 +1,9 @@
+{% load release_notes %}
+
+<a href="{% url 'download-redirect' release.version 'tarball' %}">
+  {{ release.tarball.name }}</a><br>
```
Both syntax are completely equivalent in HTML (the trailing `/` is stripped during parsing): https://developer.mozilla.org/en-US/docs/Glossary/Void_element#self-closing_tags

bmispelon commented on releases/admin.py:
```diff
@@ -1,10 +1,30 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        for fieldname, accept in _ARTIFACT_FILE_EXTENSIONS.items():
+            self.fields[fieldname].widget.attrs["accept"] = accept
```
The `accept` attribute on the `<input>` also had the benefit of filtering the list of visible files in the file picker, it's not just about validation (which should be done server side, I agree).

bmispelon commented:
I've fixed the tests and pushed a few changes:

- Removed the validators: the validation is now all handled in the `clean()` (I also added a lot more tests)
- Removed the explicit call to `full_clean()` in `save()`. This made some unrelated tests fail and is not really needed because the modelforms from the admin will go through that validation: https://github.com/django/django/blob/main/django/forms/models.py#L498

I think we should restore the `<input accepts=...>` changes for the reasons I wrote in the comment.
Same with the `{% if release.checksum %"`: personally I'd keep it.

Otherwise, I like the changes you made to keep the case of the uploaded artifact, I think it makes sense.

bmispelon commented on releases/views.py:
```diff
@@ -40,8 +40,8 @@ def index(request):
-        redirect_url = release.get_redirect_url(kind)
-    except ValueError:
+
+    if not release.is_published or not (artifact := getattr(release, kind, None)):
```
I'm not sure I see the benefit of raising a 404 if the release is not published, since anyone can use the "raw" media URL to get to the file anyway.
Won't that also prevent the releasers from testing the checksum file (if/once we make the change to using download URLs instead of direct media ones)?

nessita replied:
My thinking was that releasers would use the old/raw URL, which require knowing the name of the artifact. I know this is not that much of a secret, but figured we could make the official `download` endpoint respect the `is_published` status for consistency sake.

nessita replied:
Reverted in a separated commit for review, we can squash that in before merge.

bmispelon commented on releases/tests.py:
```diff
@@ -157,6 +144,293 @@ def test_preview(self):
+            with self.subTest(kind=kind):
+                self.assertRedirects(response, url, 301, fetch_redirect_response=False)
+
+    def test_redirect_404(self):
+        Release.objects.create(version="1.0", is_active=True)
+
+        for kind in ["tarball", "wheel", "checksum"]:
+            response = self.client.get(f"/download/1.0/{kind}/")
+            with self.subTest(kind=kind):
+                self.assertEqual(response.status_code, 404)
+
```
This test is now superseded by `test_redirect_is_not_published`

nessita replied:
Yes, you are right, removing it.

bmispelon commented on releases/tests.py:
```diff
@@ -157,6 +144,293 @@ def test_preview(self):
+            ({"date": future, "is_active": True}, False),
+            ({"date": future, "is_active": False}, False),
+        ]
+        for i, (params, expected) in enumerate(cases):
+            release = Release.objects.create(version=f"{i}.0", **params)
```
The `is_published` property should work on unsaved models, so we don't even have to specify a version I think:

```suggestion
        for params, expected in cases:
            release = Release(**params)
```

nessita replied:
I will make another assertion for an unsaved model so we test correctness in both scenarios.

bmispelon commented on releases/models.py:
```diff
@@ -149,6 +175,13 @@ class Release(models.Model):
     }
 
     version = models.CharField(max_length=16, primary_key=True)
+    is_active = models.BooleanField(
```
I think this new field should be taken into account in `ReleaseManager.active()`. Otherwise I'm not fully confident that some pages might end up listing inactive releases somehow.

nessita replied:
You are correct, I missed this manager method existed. Changing and adding tests.

nessita commented on releases/tests.py:
```diff
@@ -157,6 +144,293 @@ def test_preview(self):
+            ({"date": future, "is_active": True}, False),
+            ({"date": future, "is_active": False}, False),
+        ]
+        for i, (params, tarball_required) in enumerate(cases):
+            form = self.form_class({"version": f"{i}.0", **params})
```
Good catch! removing that part.

nessita commented:
@bmispelon I pushed an update where the `ReleaseManager.active` method was renamed to `.published`, taking into consideration the `is_active` flag. There are no other usages of `.active` in the code:
```
$ grep -nR "\.active" --exclude-dir=static_root --exclude-dir=static
blog/models.py:32:        return self.active().filter(pub_date__lte=timezone.now())
djangoproject/scss/_style.scss:377:        &.active {
djangoproject/scss/_style.scss:527:        &.active {
djangoproject/scss/_style.scss:539:        &.active {
djangoproject/scss/_style.scss:577:            &.active a {
djangoproject/scss/_style.scss:1880:    &.active {
djangoproject/scss/_style.scss:1887:            &.active {
djangoproject/scss/_style.scss:1942:        li.active .collapsing-content {
djangoproject/scss/_style.scss:3079:        &.active {
```

nessita commented on releases/models.py:
```diff
@@ -133,6 +140,26 @@ def current_version(self):
+
+def upload_to_artifact(release, filename):
+    major, minor = release.version_tuple[:2]
+    return f"releases/{major}.{minor}/{filename}"
```
YES, thanks @bmispelon :trophy:

nessita commented on releases/models.py:
```diff
@@ -172,8 +206,30 @@ class Release(models.Model):
-    is_lts = models.BooleanField("Long term support release", default=False)
+    is_lts = models.BooleanField(
+        "Long Term Support",
+        help_text='Is this an <abbr title="Long Term Support">LTS</abbr> release?',
```
Good point, will expand.

bmispelon replied:
This is the world's tiniest nitpick, and since I'm no native speaker I could be very wrong but would you actually read this as "a release for a el-tee-ess Django version"? Wouldn't you say "an el-tee-ess"?

nessita commented on docs/tests/test_models.py:
```diff
@@ -90,9 +90,11 @@ def test_most_recent_micro_release_considered(self):
+        r = Release.objects.create(version="1.8", is_active=True, date=today - 15 * day)
         d = DocumentRelease.objects.create(release=r)
-        r2 = Release.objects.create(version="1.8.1", date=today - 5 * day)
+        r2 = Release.objects.create(
+            version="1.8.1", is_active=True, date=today - 5 * day
+        )
```
Great suggestion, adding.

nessita reviewed (approved):
This looks great to me!

bmispelon commented:
Now we just need a fresh new Django version to test things out. Wanna roll out an unplanned release just for fun? 🙊
